### ISSUE TITLE
DeployStrategy is not consistent between OpenstackDataplane to OpenStackDataplaneNode

### ISSUE THREAD

**Issue by gibizer** [openstack-k8s-operators/dataplane-operator#112]
I deployed the [OpenstackDataPlane sample](https://github.com/openstack-k8s-operators/dataplane-operator/blob/main/config/samples/dataplane_v1beta1_openstackdataplane.yaml) from the repo and I got the following result:
```
$ oc get -o yaml   OpenstackDataPlane
apiVersion: v1
items:
- apiVersion: dataplane.openstack.org/v1beta1
  kind: OpenStackDataPlane
  metadata:
    annotations:
[...snip...]
  spec:
    deployStrategy:
      deploy: true
    nodes:
      edpm-compute-0:
        ansibleHost: 192.168.122.100
        deployStrategy:
          deploy: false
        hostName: edpm-compute-0
        node:
          ansibleSSHPrivateKeySecret: dataplane-ansible-ssh-private-key-secret
          ansibleVars: |
            tenant_ip: 192.168.122.100
        openStackAnsibleEERunnerImage: quay.io/openstack-k8s-operators/openstack-ansibleee-runner:latest
        role: edpm-compute
    roles:
      edpm-compute:
        deployStrategy:
          deploy: false
```
```
$ oc get -o yaml   OpenstackDataPlaneNode
apiVersion: v1
items:
- apiVersion: dataplane.openstack.org/v1beta1
  kind: OpenStackDataPlaneNode
  metadata:
    creationTimestamp: "2023-03-21T11:56:06Z"
    generation: 1
    labels:
      openstackdataplanerole: edpm-compute
    name: edpm-compute-0
    namespace: openstack
    resourceVersion: "139974"
    uid: 3fd5246e-ce4d-4397-9789-50647846ef61
  spec:
    ansibleHost: 192.168.122.100
    deployStrategy:
      deploy: false
```

The deployStrategy is different on the top level than in the lower layers.

This shows two things:
1. the deployStrategy is not propagated from OpenstackDataplane to OpenStackDataplaneNode (and probably Role)
2. the defaulting behavior of deployStrategy is broken as the original sample I applied contains
```
spec:
  deployStrategy:
      deploy: false
```
but the resulting CR contains:
```
  spec:
    deployStrategy:
      deploy: true
```

**Comment by slagle** [comment]
The DataPlane DeployStrategy.deploy won't propagate to the node, only the role. The role does a batched deploy with all the nodes in the inventory, so deploy is not set individually on each node.

**Comment by gibizer** [comment]
> so deploy is not set individually on each node.

that still feels weird if lookig at the Node.Spec it has deploy=false but the Node gets deployed.

**Comment by slagle** [comment]
> > so deploy is not set individually on each node.
> 
> that still feels weird if lookig at the Node.Spec it has deploy=false but the Node gets deployed.

We do need to update the conditions on the Node to indicate it's deploying/deployed. That's not done yet.

The "deploy" bool means "start an ansible execution with the inventory for this resource". So, we could rename it someway to make it clearer, or separate it into different properties (nodeDeploy, roleDeploy). Then we could set roleDeploy on the node to indicate a role deploy was requested, but really that is what the conditions are for.

**Comment by gibizer** [comment]
> We do need to update the conditions on the Node to indicate it's deploying/deployed. That's not done yet.
ack that will help.

> The "deploy" bool means "start an ansible execution with the inventory for this resource". So, we could rename it someway to make it clearer, or separate it into different properties (nodeDeploy, roleDeploy). Then we could set roleDeploy on the node to indicate a role deploy was requested, but really that is what the conditions are for.

Yeah we need some better naming here. In my understanding the Node.Spec describe the expected state of the resource. So if Node.Spec.DeployStrategy.Deploy == false that means this node expected not to be deployed.

**Comment by bshephar** [comment]
Ideally, I would like to remove this entirely as part of:
https://github.com/openstack-k8s-operators/dataplane-operator/pull/217

My idea would be that when a `OpenStackDataPlane` object is created, the user would have the expectation that it WOULD deploy at that stage. 

Once the successful deployment has completed, then we assert the annotation on each of the `OpenStackDataPlane` objects. 

User wishes to redeploy, they simply remove the annotation from the user owned `OpenStackDataPlane`. We remove it from the Controller managed objects and reassert the state.

Few things pending with this:
1. It's entirely conceivable that a user would like to assert the state in their environment without actually changing anything. Think for example that someone has ssh'd into a node, made some manual changes to `nova.conf` and broken things. Now they want to re-run the execution. Even though from a k8s perspective, nothing has changed about the object. We still need to provide some mechanism to re-execute the jobs. This is where I think the deployIdentifier comes in. Essentially:
- User removes annotation;
- We infer this to mean that they wish to re-run Ansible tasks. So we generate a new, random `deployIdentifier`.
- We set the `deployIdentifer` in the `status` of the user managed object, and we set it in the `spec` of the Controller managed objects which means that the k8s object has changed and now we can reconcile the change.

2. We need to make sure any logic in our operator that depends on a status condition is rock solid and as simple as we can possibly make it to ensure that we can't get stuck in some indefinite state.

Still a bit of testing and brainstorming to go. But feel free to share any thoughts on:
https://issues.redhat.com/browse/OSP-25708

And the associated Google doc.
